Fixed NaN propagation in optimal_step_size using nanmin/nanmax

SuriyaaMM opened the pull request:
[issue](https://github.com/jax-ml/jax/issues/14612) (I am not English native, sorry for my bad english)

The suggested fix in the issue seems to work. Here are my scripts and test results.
**Test Results**
```
====================
1. Testing with ORIGINAL (BROKEN) function
--- fn: optimal_step_size() ---
ORIGINAL: last_step=0.009999999776482582, mean_error_ratio=nan, dfactor_initial=0.20000000298023224
ORIGINAL: dfactor_adjusted=0.20000000298023224
ORIGINAL: candidate_val=nan
ORIGINAL: inner_max=nan
ORIGINAL: final_factor=nan
ORIGINAL: Exiting. optimal_step=nan
Got nan
====================
====================
2. Testing with ORIGINAL (BROKEN) function
--- fn: optimal_step_size() ---
ORIGINAL: last_step=0.009999999776482582, mean_error_ratio=0.5, dfactor_initial=0.20000000298023224
ORIGINAL: dfactor_adjusted=1.0
ORIGINAL: candidate_val=1.0338284969329834
ORIGINAL: inner_max=1.0338284969329834
ORIGINAL: final_factor=1.0338284969329834
ORIGINAL: Exiting. optimal_step=0.010338285006582737
Got 0.010338285006582737
====================
====================
1. Testing with FIXED function
----- fn: optimal_step_size() -----
FIXED: last_step=0.009999999776482582, mean_error_ratio=nan, dfactor_initial=0.20000000298023224
FIXED: dfactor_adjusted=0.20000000298023224
FIXED: candidate_val=nan
FIXED: inner_max=0.20000000298023224
FIXED: final_factor=0.20000000298023224
FIXED: Exiting. optimal_step=0.0020000000949949026
Got 0.0020000000949949026
====================
====================
2. Testing with FIXED function
----- fn: optimal_step_size() -----
FIXED: last_step=0.009999999776482582, mean_error_ratio=0.5, dfactor_initial=0.20000000298023224
FIXED: dfactor_adjusted=1.0
FIXED: candidate_val=1.0338284969329834
FIXED: inner_max=1.0338284969329834
FIXED: final_factor=1.0338284969329834
FIXED: Exiting. optimal_step=0.010338285006582737
Got 0.010338285006582737
====================
```
**PyTest Results** (I tested only ode_test.py)
```
============================= test session starts ==============================
platform linux -- Python 3.11.0, pytest-8.4.0, pluggy-1.6.0
rootdir: /home/mms/dev/jax
configfile: pyproject.toml
plugins: xdist-3.7.0, hypothesis-6.135.0
created: 10/10 workers
10 workers [14 items]

..............                                                           [100%]
============================== 14 passed in 8.73s ==============================
```
These are the function modifications that I did to the original optimal_step_function for testing
```
from jax.debug import print as dprint
def optimal_step_size_fixed(last_step, mean_error_ratio, safety=0.9, ifactor=10.0,
                      dfactor=0.2, order=5.0):
  """Compute optimal Runge-Kutta stepsize"""
  dprint("----- fn: optimal_step_size() -----")
  dprint("FIXED: last_step={last_step}, mean_error_ratio={mer}, dfactor_initial={dfac_i}",
         last_step=last_step, mer=mean_error_ratio, dfac_i=dfactor)

  dfactor_adjusted = jnp.where(mean_error_ratio < 1, 1.0, dfactor)
  dprint("FIXED: dfactor_adjusted={dfac_a}", dfac_a=dfactor_adjusted)

  candidate_val = mean_error_ratio**(-1.0 / order) * safety
  dprint("FIXED: candidate_val={cv}", cv=candidate_val)

  inner_max = jnp.nanmax(jnp.array([candidate_val, dfactor_adjusted]))
  dprint("FIXED: inner_max={im}", im=inner_max)

  factor_fixed = jnp.nanmin(jnp.array([ifactor, inner_max]))
  dprint("FIXED: final_factor={f}", f=factor_fixed)

  optimal_step = jnp.where(mean_error_ratio == 0, last_step * ifactor, last_step * factor_fixed)
  dprint("FIXED: Exiting. optimal_step={os}", os = optimal_step)

  return optimal_step

def optimal_step_size_original(last_step, mean_error_ratio, safety=0.9, ifactor=10.0,
                               dfactor=0.2, order=5.0):
  """Compute optimal Runge-Kutta stepsize (ORIGINAL, problematic version)."""
  dprint("--- fn: optimal_step_size() ---")
  dprint("ORIGINAL: last_step={last_step}, mean_error_ratio={mer}, dfactor_initial={dfac_i}",
         last_step=last_step, mer=mean_error_ratio, dfac_i=dfactor)

  dfactor_adjusted = jnp.where(mean_error_ratio < 1, 1.0, dfactor)
  dprint("ORIGINAL: dfactor_adjusted={dfac_a}", dfac_a=dfactor_adjusted)

  candidate_val = mean_error_ratio**(-1.0 / order) * safety
  dprint("ORIGINAL: candidate_val={cv}", cv=candidate_val)

  inner_max = jnp.maximum(candidate_val, dfactor_adjusted)
  dprint("ORIGINAL: inner_max={im}", im=inner_max)

  factor_original = jnp.minimum(ifactor, inner_max)
  dprint("ORIGINAL: final_factor={f}", f=factor_original)

  optimal_step = jnp.where(mean_error_ratio == 0, last_step * ifactor, last_step * factor_original)
  dprint("ORIGINAL: Exiting. optimal_step={os}", os=optimal_step)

  return optimal_step
```
These are the test functions that I used, However I wasn't sure of results of the ODE part, There were NAN's even though the changed function doesn't give out one. But still I am adding them here.
```
# ODE Testing part
# import jax
# import jax.numpy as jnp
# from jax.experimental import ode

# # I am assuming that large step sizes lead to nan, so  dy/dt = e^y something like that
# # would grow fast
# def diverging_ode_func(y, t):
#     return y**3
# def diverging_ode_func2(y, t):
#     return y**4
# def diverging_ode_func3(y, t):
#     return jnp.exp(y)

# # initial conditions
# y0 = jnp.array([1.0])
# # integrate from t = 0 to 1
# t = jnp.array([0.0, 1.0])

# rtol = 1e-3
# atol = 1e-3

# try:
#     # solve
#     solution = ode.odeint(diverging_ode_func, y0, t, rtol=rtol, atol=atol)
#     solution2 = ode.odeint(diverging_ode_func2, y0, t, rtol=rtol, atol=atol)
#     solution3 = ode.odeint(diverging_ode_func3, y0, t, rtol=rtol, atol=atol)

#     print("\n--- Integration Complete ---")
#     print("Final Solution Array:")
#     print(solution)
#     print("Final Solution Array: (2)")
#     print(solution2)
#     print("Final Solution Array: (3)")
#     print(solution3)

# except Exception as e:
#     print(f"\n!!! ERROR during integration: {e} !!!")


# -------------------------------------------------
# optimal_step_size testing part
import jax
import jax.numpy as jnp
from jax.experimental import ode
from pprint import pprint

class Test_optimal_step_size:
    def test_nan_propagation_bug(self):
        last_step = 0.01
        mean_error_ratio = jnp.nan
        dfactor = 0.2
        print("=" * 20)
        print("1. Testing with ORIGINAL (BROKEN) function")
        result = ode.optimal_step_size_original(last_step, mean_error_ratio, dfactor = 0.2)
        print(f"Got {result}")
        print("=" * 20)
        # some random value
        mean_error_ratio = 0.5
        print("=" * 20)
        print("2. Testing with ORIGINAL (BROKEN) function")
        result = ode.optimal_step_size_original(last_step, mean_error_ratio, dfactor = 0.2)
        print(f"Got {result}")
        print("=" * 20)
    
    def test_nan_fix_works(self):
        last_step = 0.01
        mean_error_ratio = jnp.nan
        dfactor = 0.2
        print("=" * 20)
        print("1. Testing with FIXED function")
        result = ode.optimal_step_size_fixed(last_step, mean_error_ratio, dfactor = 0.2)
        print(f"Got {result}")
        print("=" * 20)
        # some random value
        mean_error_ratio = 0.5
        print("=" * 20)
        print("2. Testing with FIXED function")
        result = ode.optimal_step_size_fixed(last_step, mean_error_ratio, dfactor = 0.2)
        print(f"Got {result}")
        print("=" * 20)

# This one is for the actual modified function
# class Test_optimal_step_size:    
#     def test_nan_fix_works(self):
#         last_step = 0.01
#         mean_error_ratio = jnp.nan
#         dfactor = 0.2
#         print("=" * 20)
#         print("1. Testing with FIXED function")
#         result = ode.optimal_step_size(last_step, mean_error_ratio, dfactor = 0.2)
#         print(f"Got {result}")
#         print("=" * 20)
#         # some random value
#         mean_error_ratio = 0.5
#         print("=" * 20)
#         print("2. Testing with FIXED function")
#         result = ode.optimal_step_size(last_step, mean_error_ratio, dfactor = 0.2)
#         print(f"Got {result}")
#         print("=" * 20)


if __name__ == "__main__":
    
    ots = Test_optimal_step_size()
    ots.test_nan_fix_works()
    ots.test_nan_propagation_bug()
```

SuriyaaMM committed:
Fix: avoid NaN propagation in optimal_step_size using nanmin/nanmax

jakevdp commented:
Hi - thanks for the report. I don't think `nanmin` and `nanmax` are the right approach here, because concatenation and reduction adds a lot of overhead when compared to the existing element-wise comparison routine.

I'd be happy to help put together an alternative, but before that could you please add a test case in `ode_test.py` that covers the case you have in mind?

SuriyaaMM commented:
Hmm, Hello, First of all I would like to clarify I don't know much about solving ODE and that kind of stuff. So please excuse me if I am wrong.
The person who raised this issue was concerned with propagation of NAN throughout his/her calculations. He also suggested a fix that could possibly fix it and that's exactly what I changed in the code. I think you can see that in the diff.

If you see here, once candidate_val is nan, everything else becomes nan.
```
====================
1. Testing with ORIGINAL (BROKEN) function
--- fn: optimal_step_size() ---
ORIGINAL: last_step=0.009999999776482582, mean_error_ratio=nan, dfactor_initial=0.20000000298023224
ORIGINAL: dfactor_adjusted=0.20000000298023224
ORIGINAL: candidate_val=nan
ORIGINAL: inner_max=nan
ORIGINAL: final_factor=nan
ORIGINAL: Exiting. optimal_step=nan
Got nan
====================
```
but here, it replaces that with dfactor_adjusted
```
====================
1. Testing with FIXED function
----- fn: optimal_step_size() -----
FIXED: last_step=0.009999999776482582, mean_error_ratio=nan, dfactor_initial=0.20000000298023224
FIXED: dfactor_adjusted=0.20000000298023224
FIXED: candidate_val=nan
FIXED: inner_max=0.20000000298023224
FIXED: final_factor=0.20000000298023224
FIXED: Exiting. optimal_step=0.0020000000949949026
Got 0.0020000000949949026
====================
```

Now I am not 100% sure whether to use dfactor_adjusted when it becomes NAN, but other libraries like scipy, in their implementation they are using a min_step like [scipy](https://github.com/scipy/scipy/blob/main/scipy/integrate/_ivp/rk.py)
```
min_step = 10 * np.abs(np.nextafter(t, self.direction * np.inf) - t)
```

Also I don't have any specific case or something in my head. The only case I can think of is the one given above.
